Objective-C/CBLProperties.mm

```diff
@@ -22,85 +22,83 @@
 
 
 @implementation CBLProperties {
-    FLDict _root;             // For CBLDocument
-    NSDictionary* _rootProps; // For CBLSubdocument
-    
+    FLDict _root;
     NSMutableDictionary* _properties;
+    NSMutableSet* _changedNames;
     BOOL _hasChanges;
 }
 
+@synthesize root=_root, hasChanges=_hasChanges;
+
 
-@synthesize hasChanges=_hasChanges;
+#pragma mark - PROPERTY ACCESSORS
 
 
 - (nullable NSDictionary*) properties {
-    if (!_properties)
-        _properties = [self.savedProperties mutableCopy];
-    return _properties;
+    return [self currentPropertiesAsJson: NO];
 }
 
 
 - (void) setProperties: (NSDictionary*)properties {
-    _properties = properties ? [properties mutableCopy] : [NSMutableDictionary dictionary];
-    [self markChanges];
+    _properties = [self convertProperties: properties];
+    _changedNames = nil;
+    [self markChangedKeys: [_properties allKeys]];
```

Comment:
I don't think this will do the right thing for keys that are in `_root` but are not in the new `properties`. They should be added to `_changedNames`, but they won't be.